src/ctap/mod.rs

```diff
@@ -1399,19 +1399,14 @@ impl CtapState {
         params: AuthenticatorVendorUpgradeParameters,
     ) -> Result<ResponseData, Ctap2StatusCode> {
         let AuthenticatorVendorUpgradeParameters { offset, data, hash } = params;
-        let upgrade_locations = env
-            .upgrade_storage()
-            .ok_or(Ctap2StatusCode::CTAP1_ERR_INVALID_COMMAND)?;
-        upgrade_locations
-            .write_partition(offset, &data)
-            .map_err(|_| Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER)?;
-        let written_slice = upgrade_locations
-            .read_partition(offset, data.len())
-            .map_err(|_| Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER)?;
-        let written_hash = Sha256::hash(written_slice);
-        if hash != written_hash {
+        let calculated_hash = Sha256::hash(&data);
+        if hash != calculated_hash {
             return Err(Ctap2StatusCode::CTAP2_ERR_INTEGRITY_FAILURE);
         }
+        env.upgrade_storage()
+            .ok_or(Ctap2StatusCode::CTAP1_ERR_INVALID_COMMAND)?
+            .write_bundle(offset, data)
```

Comment:
I still like the idea of checking the hash after it has been written in flash in case corruption happens during the write. What do you think of pulling the hash calculation into `write_bundle`?

Reply:
That's how it was built before, and we decided to separate the checks instead:
https://github.com/google/OpenSK/pull/551#discussion_r984737093

What we do now is check the hash in CTAP, and compare the slices in Env. Would that work?

Reply:
Yeah, that works for me